### ISSUE TITLE
Proposal: supporting second derivatives

### ISSUE THREAD

**Issue by davidgonmar** [tinygrad/tinygrad#5858]
Second derivatives are useful on some cases (see https://github.com/tinygrad/tinygrad/pull/5701), but not possible by the way autodiff works in Tinygrad. Here is my take on how they could be supported with no expected extra cost:

Let's take scalar derivatives as an example (higher dimensional derivatives are pretty much the same in the implementation context).
A simple example is for torch is:
```python
def f(x): return x**3
t = Tensor(3.0, requires_grad=True)
fout = f(t) # Tensor(27.0)
fout.backward(retain_graph=True)
df_dt = t.grad # the derivative function is 3x ** 2, so derivative will be 81.0. This means 'how does the output of the function change when I change t
```
What happens here is that fout 'remembers' where it came from w.r.t its inputs (the leaf tensors marked as requires_grad). Now, if we wanted to do a second derivative in PyTorch, it would be like this
```python
t.grad = None # the graph will be 'backpropagated again', but we don't want it to affect the new calculation.
df_dt.backward() # second derivative is 6x, so derivative will be 18.0. This means 'how does the first derivative change when I change t'. This is why we are doing .backward() on the first derivative.
d2f_dt2 = t.grad
```
Now, what is happening here? df_dt knows where it came from w.r.t its inputs (t in this case)! (at least that is what happens in torch). However, this is how a derivative is created in tinygrad:
```python
class Neg(Function):
  def forward(self, x:LazyBuffer) -> LazyBuffer: return x.e(UnaryOps.NEG)
  def backward(self, grad_output:LazyBuffer) -> LazyBuffer: return grad_output.e(UnaryOps.NEG)
```
As you can see, the .backward() works with LazyBuffers, but they do not handle autodiff. So if we run the previous example in tinygrad, the second derivative will come out as 'None', because the first derivative does not know where it came from w.r.t its inputs (the _ctx of the grad is None). The solution here is, at first glance:
1 - Support retain_graph : easy, just do not delete the self._ctx if retain_graph = True. That way the graph information remains.
2 - The Function class needs to work at Tensor level for the backward function. Example:
```python
class Neg(Function):
  def forward(self, x:LazyBuffer) -> LazyBuffer: return x.e(UnaryOps.NEG)
  def backward(self, grad_output:Tensor) -> Tensor: return -grad_output
```
This way, the gradients know where they came from at Tensor level, so you can compute 'how does the gradient change wrt this input'. In a similar way that a computational graph is constructed from inputs to outputs, one will be constructed from inputs to grads. This should not change performance at all (probably) since the underlying LazyBuffer ops should be similar. Lines will not be more (or at least significantly more).
If this looks nice, I can try to get it working:) Just wanted to ask for opinions since it is a significant change.

**Comment by geohot** [comment]
If you want to change Backward to be tensor, what do you do about saved parameters?

**Comment by davidgonmar** [comment]
> If you want to change Backward to be tensor, what do you do about saved parameters?

I don't get this. Why would it interfer with saved params?

**Comment by geohot** [comment]
saved params are LazyBuffers. look at the other functions.

better yet, just try this and see if it works

**Comment by davidgonmar** [comment]
> saved params are LazyBuffers. look at the other functions.
> 
> better yet, just try this and see if it works

Oh got the quesrion, did not understand it at first. Well, the only way is to save them also as tensors. This is what torch does under the hood. Will look into the simplest way to do it.

**Comment by chenyuxyz** [comment]
supported with the new gradient api
